Add data sharing consent redirect for more course tabs

andy-armstrong opened the pull request:
## [LEARNER-394](https://openedx.atlassian.net/browse/LEARNER-394)

### Description

This change updates the ``CourseTabView`` class to implement the data sharing consent redirect, which means that all of its subclasses will have the correct behavior. This currently is just the new course home experience (which hasn't launched yet), and the "Discussion" tab, but the intention is that eventually all tabs will be implemented as fragments which render with this view.

As part of this change, I converted all views to use the decorator to provide the redirect, rather than directly implementing the redirect for themselves. I also refactored all of the enterprise helper code into a new enterprise_support Django app to contain all the logic in one place.

### Sandbox

I've updated my sandbox to have this code. @haikuginger, can you suggest how it can be tested.

https://andy-armstrong.sandbox.edx.org/

### Testing
- [ ] Unit, integration, acceptance tests as appropriate

FYI: @nasthagiri @dianakhuang 

### Post-review
- [ ] Rebase and squash commits

andy-armstrong commented on common/djangoapps/util/enterprise_helpers.py:
```diff
@@ -203,6 +203,12 @@ def inner(request, course_id, *args, **kwargs):
+                user.get('real_user', user),
+                course_id,
+            )
             return redirect(consent_url)
```
Reviewers: this decorator always redirects to the same URL after the user grants consent, while some of the custom implementations redirect back to the originating page. Should this method be changed to always go back to the URL that the user originally requested?

andy-armstrong replied:
I looked at this some more, and the default is exactly as I described (it will redirect back to the original path). So I think this is all set.

andy-armstrong commented:
jenkins run js

andy-armstrong commented on lms/djangoapps/discussion/tests/test_views.py:
```diff
@@ -363,14 +363,12 @@ def setUp(self):
         (ModuleStoreEnum.Type.split, False, 1, 3, 3, 12, 1),
         (ModuleStoreEnum.Type.split, False, 50, 3, 3, 12, 1),
 
-        # Enabling Enterprise integration increases the number of (cached and uncached) SQL queries by 1,
```
Reviewers: this code no longer has an extra query because we don't do the user consent check in JSON calls. It only makes sense to redirect HTML requests which are now handled in the ``CourseTabView``.

haikuginger commented:
For testing:

1. Using the Django admin, create an EnterpriseCustomer.
2. Check the "Enable data sharing consent" box for that EnterpriseCustomer.
3. Go to the Manage Learners view within that EnterpriseCustomer's admin (look at the top-right of the main edit view for the button).
4. Enter the email address of your user on the right-hand side where it prompts for an email. Also enter a course ID present on your sandbox, select a valid course mode, and click the Submit button to simultaneously link your user to the Enterprise and enroll it in that course.
5. Go to the Dashboard, and attempt to browse to the course; observe that you're blocked from browsing to it, and that you're redirected to a page prompting for data sharing consent (note that if your sandbox is older, you may need to rebuild your static assets, as this view changed recently).
6. Decline consent and observe that you're redirected to the Dashboard.
7. Attempt to manually browse to the URLs for the individual tabs of the course, and observe that each one is similarly blocked.

haikuginger commented:
@andy-armstrong, I haven't done a thorough review yet - it looks like you're moving some stuff around (which I'm not opposed to), but you may have missed this template, which is a recent merge:

https://github.com/edx/edx-platform/blob/master/common/templates/util/enterprise_consent_declined_notification.html

It's referred to in the Dashboard template (and in the edX.org themed Dashboard template).

andy-armstrong commented:
@haikuginger thanks for the great feedback. I had indeed missed the existence of that template. Thinking it through further, it then seemed clear that the Enterprise code needed to be a Django app so that it can provide templates. After discussion with @robrap, we've decided to follow the recommendation of the ["Taming the Monolith"](https://openedx.atlassian.net/wiki/display/TNL/Taming+the+Monolith) project and create a new enterprise_support Django app in the [openedx/features](https://github.com/edx/edx-platform/tree/master/openedx/features) directory. This directory now contains the api.py file, as well as templates and tests. It will also make it simpler in the future to ensure that all enterprise extensions live in one location (if they can't go into the edx-enterprise library for some reason).

To try to clarify this, I've added a README to the root of the Django app:

https://github.com/edx/edx-platform/tree/4efc49d7054ee4c668872ba1f9c4842b3ba3a099/openedx/features/enterprise_support

I've refactored the code into this proposed Django app so let me know what you think. @mattdrayer, I'd love your opinion too.

robrap commented:
jenkins run js

andy-armstrong commented:
jenkins run python

andy-armstrong commented:
@haikuginger @mattdrayer @robrap I have the sandbox working correctly for the redirect using the test user "EnterpriseUser" (standard password). You can see the behavior here for the new course outline:

https://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+AAPTC+1/course

andy-armstrong commented:
jenkins run js

haikuginger commented on openedx/features/enterprise_support/README.rst:
```diff
@@ -0,0 +1,13 @@
+
+Note: this Django app is only necessary due to limitations in extending
+edx-platform through installed libraries. The eventual goal is to have
+all of this functionality provided through edx-enterprise. New code should
```
Not exactly. Our huge, big, long-term goal is to actually implement Enterprise as an IDA - in which case, the stuff in Enterprise Support would likely just be REST clients for various tasks. We've been intentionally abstracting calls into edx-enterprise at this choke point so that when (if) that eventually happens, we have a single point that we can refactor without affecting the rest of the platform.

andy-armstrong replied:
@haikuginger thanks, that's helpful. IMO it would still be the case that this client would live in a separate repo, and it would be pip installed into edx-platform. We are trying to get to the point where edx-platform does not need to be modified in order to plug in new capabilities.

Shall I just remove this paragraph for now, and then you can add in any details you'd like to see here?

haikuginger replied:
@andy-armstrong, that sounds like a good idea.

andy-armstrong commented:
jenkins run js

andy-armstrong commented:
jenkins run python

andy-armstrong commented:
jenkins run js

andy-armstrong commented:
jenkins run bokchoy

andy-armstrong commented:
I'm going to merge since the same ten failures are happening on master. I'm then going to see if I can work out what is going on there.
